Singer.API/Controllers/DataControllerBase.cs

```diff
@@ -0,0 +1,180 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
+using Singer.DTOs;
+using Singer.Helpers;
+using Singer.Helpers.Exceptions;
+using Singer.Services.Interfaces;
+using System;
+using System.ComponentModel;
+using System.Threading.Tasks;
+
+namespace Singer.Controllers
+{
+   /// <summary>
+   /// Abstract class that implements the basic CRUD operations for a controller.
+   /// </summary>
+   /// <typeparam name="TEntity">The type of the entity to manipulate in the database.</typeparam>
+   /// <typeparam name="TDTO">The type that will be exposed to the outside world.</typeparam>
+   [Route("api/[controller]")]
+   public abstract class DataControllerBase<TEntity, TDTO> : Controller
+      where TEntity : class, IIdentifiable
+      where TDTO : class
+   {
+      #region CONSTRUCTORS
+
+      /// <summary>
+      /// Constructs a new instance of the <see cref="DataControllerBase{TEntity, TDTO}"/> class.
+      /// </summary>
+      /// <param name="databaseService">Service to perform operations on the database.</param>
+      protected DataControllerBase(IDatabaseService<TEntity, TDTO> databaseService)
+      {
+         DatabaseService = databaseService;
+      }
+
+      #endregion CONSTRUCTORS
+
+
+      #region PROPERTIES
+
+      /// <summary>
+      /// Service to perform operations on the database.
+      /// </summary>
+      protected IDatabaseService<TEntity, TDTO> DatabaseService { get; }
+
+      #endregion PROPERTIES
+
+
+      #region METHODS
+
+      #region post
+
+      /// <summary>
+      /// The POST method for the controller to create a new entity in the database.
+      /// </summary>
+      /// <param name="dto">The new element to create in the database.</param>
+      /// <returns>The new created entity.</returns>
+      [HttpPost]
+      [ProducesResponseType(StatusCodes.Status201Created)]
+      [ProducesResponseType(StatusCodes.Status500InternalServerError)]
+      public async Task<ActionResult<TDTO>> Create(TDTO dto)
+      {
+         var returnUser = await DatabaseService.CreateAsync(dto);
```

Comment:
Should be `returnItem` or similar? This is not specific for users

Reply:
fixin it